lib/telemetry_metrics.ex

```diff
@@ -1,5 +1,338 @@
+  every flush and can be decremented, whereas counters in Prometheus are monotonically increasing.
+  `Telemetry.Metrics` doesn't focus on those details - instead, it describes what the end user,
+  operator, expects to see when using the metric of particular type. This implies that in most
+  cases aggregated metrics won't be visible inside the BEAM, but in exchange aggregations can be
+  implemented in a way that makes most sense for particular system. Finally, one could also
+  implement an in-VM "reporter" which would aggregate the metrics and expose them inside the BEAM;
+  When there is a need to swap the reporters, and if both reporters are following the metric types
+  specification, then the end result of aggregation
+
+  ### Requirements for reporters
+
+  Reporters should accept metric specifications and subscribe to relevant events. When those events
+  are emitted, metric should be updated (either in-memory or by contacting external system) in such
+  a way that the user is able to view metric values as described in the "Metric types" section.
+
+  If the reporter does not support the metric given to it, it should log a warning.
+
+  Reporters should also document how `Telemetry.Metrics` metric types, names tags are translated to
+  metric types and identifiers in the system they publish metrics to.
   """
+
+  @type metric_name :: [atom(), ...]
+  @type metric_type :: :counter | :sum | :last_value | :distribution
+  @type tags_fun :: (Telemetry.event_metadata() -> %{tag_key => term()})
+  @type tag_key :: atom()
+  @type description :: String.t()
+  @type unit :: atom()
+  @type counter_options :: [metric_option()]
+  @type sum_options :: [metric_option()]
+  @type last_value_options :: [metric_option()]
+  @type metric_option ::
+          {:name, metric_name()}
+          | {:tags_fun, tags_fun() | nil}
+          | {:tag_keys, [tag_key()]}
+          | {:description, description()}
+          | {:unit, unit()}
+
+  defmodule Metric do
+    @moduledoc """
+    A metric specification.
+
+    This struct should be used by reporters to initialize the metric at runtime.
+    """
+
+    alias Telemetry.Metrics
+
+    defstruct [:name, :type, :event_name, :tags_fun, :tag_keys, :description, :unit]
+
+    @type t :: %__MODULE__{
+            name: Metrics.metric_name(),
+            type: Metrics.metric_type(),
+            event_name: Telemetry.event_name(),
+            tags_fun: Telemetry.Metrics.tags_fun() | nil,
+            tag_keys: [Metrics.tag_key()],
+            description: Metrics.description(),
+            unit: Metrics.unit()
+          }
+  end
+
+  # API
+
+  @doc """
+  Returns a specification of counter metric.
+
+  See "Metric specifications" section in the top-level documentation of this module for more
+  information.
+  """
+  @spec counter(Telemetry.event_name(), counter_options()) :: Metric.t()
+  def counter(event_name, options) do
```

Comment:
I have sketched a bit this API and I think we should allow, in this case, the event name to also be a string:

    counter("my_app.requests", ...)

is quite more readable and, opposite to the `telemetry` core, the costs of converting a string to a list of atoms at this moment is very cheap. Thoughts?

Reply:
I understand the argument for readability, but I'm concerned with different representation of event and metric. There will be at least one conversion, between Telemetry metric name and reporter metric name. If we allow this, then the metric name will need to be converted in two places. I'm not sure if this won't be confusing, what do you think?

Reply:
We usually raise nice error messages on bad event names, so I believe we can tackle it from that front.

Btw, we will normalize the event and metric name before putting them into the metric structs, so this should be only on the user facing API, which would benefit from this change the most.

Reply:
Okay, let's give it a try.